ddtrace/tracer/textmap.go

```diff
@@ -487,3 +488,125 @@ func (*propagatorB3) extractTextMap(reader TextMapReader) (ddtrace.SpanContext,
+// The current specification only supports a single flag called sampled,
+// which is equal to 00000001 when no other flag is present.
+// tracestateHeader is a comma-separated list of list-members with a <key>=<value> format,
+// where each list-member is managed by a vendor or instrumentation library.
+func (*propagatorW3c) injectTextMap(spanCtx ddtrace.SpanContext, writer TextMapWriter) error {
+	ctx, ok := spanCtx.(*spanContext)
+	if !ok || ctx.traceID == 0 || ctx.spanID == 0 {
+		return ErrInvalidSpanContext
+	}
+	flags := ""
+	p, ok := ctx.samplingPriority()
+	if ok && p >= ext.PriorityAutoKeep {
+		flags = "01"
+	} else {
+		flags = "00"
+	}
+
+	writer.Set(traceparentHeader, fmt.Sprintf("00-%032x-%016x-%v", ctx.traceID, ctx.spanID, flags))
+	// if context priority / origin / tags were updated after extraction,
+	// or the tracestateHeader doesn't start with `dd=`
+	// we need to recreate tracestate
+	if ctx.updated ||
+		(ctx.trace != nil && ctx.trace.propagatingTags != nil && !strings.HasPrefix(ctx.trace.propagatingTags[tracestateHeader], "dd=")) {
+		writer.Set(tracestateHeader, composeTracestate(ctx, p, ctx.trace.propagatingTags[tracestateHeader]))
+	} else {
+		writer.Set(tracestateHeader, ctx.trace.propagatingTags[tracestateHeader])
+	}
+	return nil
+}
+
+// composeTracestate creates a tracestateHeader from the spancontext.
+// The Datadog tracing library is only responsible for managing the list member with key dd,
+// which holds the values of the sampling decision(`s:<value>`), origin(`o:<origin>`),
+// and propagated tags prefixed with `t.`(e.g. _dd.p.usr.id:usr_id tag will become `t.usr.id:usr_id`).
+// All tag keys in the list must have all invalid characters replaced with the underscore.
+// Invalid key characters include characters outside the ASCII range 0x20 to 0x7E, space, comma and equal sign.
+// All tag values in the list must have all invalid characters replaced with the underscore.
+// Invalid value characters include characters outside the ASCII range 0x20 to 0x7E, space,
+// comma(reserved for tracestate list-member separator), semi-colon(reserved for separator between entries),
+// and tilde sign(reserved to represent equals sign).
+func composeTracestate(ctx *spanContext, priority int, oldState string) string {
+	keyRgx := regexp.MustCompile(",|=|[^\\x20-\\x7E]+")
+	valueRgx := regexp.MustCompile(",|;|:|[^\\x20-\\x7E]+")
+
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("dd=s:%d;", priority))
+	listLength := 1
+
+	if ctx.origin != "" {
+		b.WriteString(fmt.Sprintf("o:%s",
+			valueRgx.ReplaceAllString(ctx.origin, "_")))
+		listLength++
+	}
+
+	for k, v := range ctx.trace.propagatingTags {
+		if !strings.HasPrefix(k, "_dd.p.") {
+			continue
+		}
+		// Datadog propagating tags must be appended to the tracestateHeader
+		// with the `t.` prefix. Tag value must have all `=` signs replaced with a tilde (`~`).
+		tag := fmt.Sprintf("t.%s:%s",
+			keyRgx.ReplaceAllString(k[6:], "_"),
```

Comment:
```suggestion
			keyRgx.ReplaceAllString(k[len("_dd.p."):], "_"),
```
Again this is both a readability measure and a safeguard